docs/assumptions.rst

```diff
@@ -1,63 +1,159 @@
+      - Using (HDF5_) and other parallel I/O libraries. Most parallel I/O libraries
+        hide the synchronization requirements of file systems from their users.
+        For example, HDF5 implements the synchronization required by the MPI-IO
+        interface so users of HDF5 do not need to perform any synchronization
+        operations explicitly in their codes.
+      - With explicit synchronization. If an application does not use a compliant
+        parallel I/O library or if the developer wishes to perform explicit
+        synchronization, the synchronization can be achieved through adding
+        explicit "flush" operations with calls to fflush(), close(), or fsync()
+        in the application source code,
+        or by supplying the "write_sync" configuration parameter to UnifyFS
+        on startup, which will cause an implicit "flush" operation after
+        every write (note: the "write_sync" mode can significantly slow down
+        write performance.).
+
+During a write phase, a process can deviate from the bulk synchronous
+I/O pattern and read any byte in
+a file, including remote data that has been written by processes
+executing on remote compute nodes in the job.
+However, the performance will differ based on which process wrote the data that
+is being read:
+      - If the bytes being read were written by the same process that is reading
+        the bytes, UnifyFS offers the fastest performance and no synchronization
+        operations are needed. This kind of access is typical in some I/O
+        libraries, e.g., HDF5, where file metadata may be updated and read by
+        the same process.
+      - If the bytes being read were written by a process executing on the same compute
+        node as the reading process, UnifyFS can offer slightly slower performance
+        than the first case and requires no additional synchronization operations.
+      - If the bytes being read were written by a process executing on a different
+        compute node than the reading process, then the performance is slower
+        than the first two cases and the application must
+        introduce synchronization operations to ensure that the most recent
+        data is read.
+In summary, reading the local data (which has been written by processes
+executing on the same compute node) will always be faster than reading
+remote data.
+
+Note that commit semantics also require synchronization for potentially conflicting
+write accesses. If an application does not enforce sequential ordering of file
+modifications during a write phase, e.g., with MPI synchronization,
+and multiple processes write concurrently to the same file offset or to an
+overlapping region, the result is undefined and may
+reflect the result of any of the processes' operations to that offset or region.
+**I don't think this paragraph is true. I think we won't return the last write even if synchronization is applied**
```

Comment:
I think the paragraph is mostly right in that concurrent writes to the same offset/region will result in undefined behavior, but perhaps it's not as clear as it could be. The problem with the current discussion is we're overloading "synchronization operations" to mean two things that are not equivalent. One type is local synchronization (i.e., flush extents to the server), the other type is global synchronization across app processes (which must be provided by something like MPI since Unify does not provide anything for that). For a consistent distributed sequential ordering of conflicting writes, there must be global sync ops between each of the local sync ops.

Reply:
I added a discussion of local and global synchronization near the top of this section. I think it helps, thanks for suggesting it. 

I still wonder about whether the most recent update will always be returned, even if correct global synchronization is used. Perhaps, though, this was a problem with MDHIM and is fixed with our new approach? I think with MDHIM we ended up with multiple key,value pairs for the same offset and we didn't know which one was most recent so one would be returned (but perhaps not the most recent). Perhaps now with our new approach, there is just one location where data for an offset is stored and when a sync operation is applied, that location is updated with the most recent data. Is that right?

Reply:
For "global sync", the word "global" is a bit strong.  What really matters is "inter-process sync", just the subset of processes that care about some shared offset, which might just be two procs out of many.  It's more like "file system synchronization" and "process synchronization".

Reply:
That's right, we don't have the problem with two key/val pairs existing for a single file offset anymore.  A single owner server is responsible for updating the main extent tree for a file.  Old entries are replaced by new entries.

Another thing regarding "undefined".  From the user's point of view, if two procs write to the same region, the result could be some strange mix of bytes from the two, even if each process issued one write.  For example, consider that two procs each do a single large write of say 1GB to the same region of the file.  Internally, each single large write may be broken into a bunch of smaller chunks/extents, which are sent to the owner server maintaining the extent tree in numerous rpc calls.  Those then get played on the server in random order.  The result is that we might get a mix of data from the two procs for that region.  It's not as if the single large write from one of the two procs wins and the other loses.  Our current implementation might actually do better than what I'm describing, but I think it'd be good to stress the "undefined" bit until users complain since it gives us some flexibility.